File: libs/HandySuites.Domain/Entities/Promocion.cs
using System.ComponentModel.DataAnnotations.Schema;
using HandySuites.Domain.Common;

namespace HandySuites.Domain.Entities;

[Table("Promociones")]
public class Promocion : AuditableEntity
{
    [Column("id")]
    public int Id { get; set; }
    [Column("tenant_id")]
    public int TenantId { get; set; }
    [Column("nombre")]
    public string Nombre { get; set; } = string.Empty;
    [Column("descripcion")]
    public string Descripcion { get; set; } = string.Empty;
    [Column("descuento_porcentaje")]
    public decimal DescuentoPorcentaje { get; set; }
    [Column("fecha_inicio")]
    public DateTime FechaInicio { get; set; }
    [Column("fecha_fin")]
    public DateTime FechaFin { get; set; }

    /// <summary>
    /// Tipo de promoción. Porcentaje (default, comportamiento legacy) o Regalo
    /// (BOGO acumulativo). Cuando Regalo, los campos CantidadCompra/CantidadBonificada
    /// dictan la regla y DescuentoPorcentaje queda en 0 e ignorado.
    /// </summary>
    [Column("tipo_promocion")]
    public TipoPromocion TipoPromocion { get; set; } = TipoPromocion.Porcentaje;

    /// <summary>Cantidad de unidades a comprar para gatillar el regalo (ej: 10).</summary>
    [Column("cantidad_compra")]
    public decimal? CantidadCompra { get; set; }

    /// <summary>Cantidad de unidades regaladas por cada CantidadCompra (ej: 1).</summary>
    [Column("cantidad_bonificada")]
    public decimal? CantidadBonificada { get; set; }

    /// <summary>
    /// FK al producto bonificado. NULL = se regala el mismo producto que se compra.
    /// Si != null se bonifica un producto distinto (ej: compra 10 X regala 1 Y).
    /// </summary>
    [Column("producto_bonificado_id")]
    public int? ProductoBonificadoId { get; set; }

    [ForeignKey(nameof(ProductoBonificadoId))]
    public Producto? ProductoBonificado { get; set; }

    public List<PromocionProducto> PromocionProductos { get; set; } = new();
    public Tenant Tenant { get; set; } = null!;
}

File: apps/mobile-app/src/db/models/Promocion.ts
import { Model } from '@nozbe/watermelondb';
import { field, text, date, readonly } from '@nozbe/watermelondb/decorators';

export default class Promocion extends Model {
  static table = 'promociones';

  @field('server_id') serverId!: number;
  @text('nombre') nombre!: string;
  @field('descuento_porcentaje') descuentoPorcentaje!: number;
  @date('fecha_inicio') fechaInicio!: Date;
  @date('fecha_fin') fechaFin!: Date;
  @text('producto_ids') productoIdsJson!: string; // JSON array: "[1,2,3]"
  @field('activo') activo!: boolean;
  @field('version') version!: number;
  @readonly @date('created_at') createdAt!: Date;
  @date('updated_at') updatedAt!: Date;
  // v18 BOGO
  @field('tipo_promocion') tipoPromocion!: number; // 0=Porcentaje, 1=Regalo
  @field('cantidad_compra') cantidadCompra!: number | null;
  @field('cantidad_bonificada') cantidadBonificada!: number | null;
  @field('producto_bonificado_id') productoBonificadoId!: number | null;

  get productoIds(): number[] {
    try { return JSON.parse(this.productoIdsJson || '[]'); }
    catch { return []; }
  }

  /** True si la promo es BOGO (compra N regala M). */
  get esRegalo(): boolean {
    return this.tipoPromocion === 1;
  }
}
